include/engine/api/base_parameters.hpp

```diff
@@ -77,7 +77,7 @@ struct BaseParameters
     };
 
     std::vector<util::Coordinate> coordinates;
-    std::vector<boost::optional<Hint>> hints;
+    std::vector<std::vector<Hint>> hints;
```

Comment:
@mjjbell 
It is a breaking change you mentioned, right? Is it possible to support both old and new version of API? I mean do not touch existing field and just add `std::vector<std::vector<Hint>> extended_hints;`? I would even make it `std::vector<WaypointHints> extended_hints` and put this vector of hints to `WaypointHints` structure to be able to extend this structure in the future if needed without breaking changes.

Reply:
:+1: I like this idea. This would make it backwards compatible with existing API users getting the existing (sub-optimal) behaviour, new users get the improved behaviour. Most people who use the HTTP API (such as osrm-frontend) will get the improvement for free.